### ISSUE TITLE
Macos Sierra: Foreign/R.chi not found

### ISSUE THREAD

**Issue by almays** [tweag/HaskellR#275]
With all dependencies installed with `brew` I run `stack build`:

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Configuring inline-r-0.9.0.0...
    Preprocessing library inline-r-0.9.0.0...
    c2hs: Foreign/R.chi not found in:
    .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build`

As soon as it says about `R` I checked `libR.pc` as said in https://github.com/tweag/HaskellR/issues/200#issuecomment-143566822.
`libR.pc` is existing.

**Comment by almays** [comment]
That appears only on subsequent builds.

In the first build appears:

    --  While building package inline-r-0.8.0.1 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/HaskellR/.stack-work/logs/inline-r-0.8.0.1.log

    Configuring inline-r-0.8.0.1...
    Preprocessing library inline-r-0.8.0.1...
    c2hs: C header contains errors:

    /usr/include/stdlib.h:139: (column 38) [ERROR]  >>> Syntax error !
      The symbol `__compar' does not fit here.

On the second attempt:

    --  While building package inline-r-0.8.0.1 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/HaskellR/.stack-work/logs/inline-r-0.8.0.1.log

    Preprocessing library inline-r-0.8.0.1...
    c2hs: C header contains errors:

    /usr/include/stdio.h:133: (column 19) [ERROR]  >>> Syntax error !
      The symbol `_close' does not fit here.

And afterwards on all subsequent builds the one in the initial post.

**Comment by qnikst** [comment]
Seems some problem with `c2hs` and system headers. Can you tell what is the version of `c2hs` that you are using?

**Comment by almays** [comment]
Well. It is `c2hs`, that is built by `stack` in the project directory.
Which is... `c2hs-0.28.1`.

It comes from the project's default `stack.yaml` with LTS6.12. I was also trying to `stack init --force`. It leads to LTS6.23 (I am not mistaken). But that changes nothing. Therefore I am just trying it again and again with the default build plan.

**Comment by almays** [comment]
Now `stack init --force` resolves to LTS6.24. `c2hs` is latest version 0.28.1. Still no success with building on macOS.

**Comment by qnikst** [comment]
Currently travis build uses OSX and passes (I'll retrigger that for convenience). It uses following commands to setup environment:
```
  - brew update
  - brew tap homebrew/science
  - brew install r zeromq python
  - pip install ipython
  - stack setup
  - stack --no-terminal -j2 build --only-snapshot --prefetch --test --bench
```
If you don't need ihaskell, then you can omit installing of `zeromq`, `python` and `ipython`.
Maybe there is some difference with your setup?

**Comment by almays** [comment]
In this way there are no errors, but also `inline-r` doesn't get built.

In the very beginning of `stack` output there is the following disclaimer:

    No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
    Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
    This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595

`stack build hsc2hs` creates executable in `.stack-work`, but still `inline-r` is not built. It seems to be `stack` issue.

**Comment by mboes** [comment]
> In this way there are no errors, but also inline-r doesn't get built.

Are you sure? I don't have OS X handy right now to check, but I see this in the Travis CI OS X logs:

```
$ stack --no-terminal -j2 build --pedantic
No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595
inline-r-0.9.0.0: configure
inline-r-0.9.0.0: build
inline-r-0.9.0.0: copy/register
H-0.9.0.1: configure
H-0.9.0.1: build
HaskellR-examples-0.1.0.0: configure
HaskellR-examples-0.1.0.0: build
H-0.9.0.1: copy/register
ihaskell-inline-r-0.1.1.0: configure
ihaskell-inline-r-0.1.1.0: build
ihaskell-inline-r-0.1.1.0: copy/register
HaskellR-examples-0.1.0.0: copy/register
Completed 4 action(s).
```

https://travis-ci.org/tweag/HaskellR/builds/173831216

**Comment by almays** [comment]
Hmmm...
With `--pedantic` key it gives me the same as simple `stack build`:

    Dahan-Macbook:HaskellR dahan$ stack --no-terminal -j2 build --pedantic
    No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
    Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
    This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595

    inline-r-0.9.0.0: configure
    inline-r-0.9.0.0: build

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/Dev/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Configuring inline-r-0.9.0.0...
    Preprocessing library inline-r-0.9.0.0...
    c2hs: C header contains errors:

    /usr/include/stdlib.h:139: (column 38) [ERROR]  >>> Syntax error !
      The symbol `__compar' does not fit here.
    Dahan-Macbook:HaskellR dahan$ stack --no-terminal -j2 build --pedantic
    No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
    Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
    This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595

    inline-r-0.9.0.0: build

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/Dev/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Preprocessing library inline-r-0.9.0.0...
    c2hs: C header contains errors:

    /usr/include/stdio.h:133: (column 19) [ERROR]  >>> Syntax error !
      The symbol `_close' does not fit here.
    Dahan-Macbook:HaskellR dahan$ stack --no-terminal -j2 build --pedantic
    No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
    Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
    This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595

    inline-r-0.9.0.0: build

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/Dev/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Preprocessing library inline-r-0.9.0.0...
    c2hs: Foreign/R.chi not found in:
    .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build

And with `stack --no-terminal -j2 build --only-snapshot --prefetch --test --bench` there are no errors, but also no build for `inline-r`.

I am not sure. Maybe I am looking for the build in the wrong place. Where it should appear in `.stack-work`?

**Comment by qnikst** [comment]
```
/usr/include/stdio.h:133: (column 19) [ERROR]  >>> Syntax error !
  The symbol `_close' does not fit here.
```

looks like a clue to this problem, for some reason `c2hs` doesn't like this header file transitively included by R headers.

Seems this is an issue  about this problem  https://github.com/haskell/c2hs/issues/85 and https://github.com/haskell/c2hs/issues/33 . 

33: suggest installing gcc instead of llvm 
and 85 suggests adding `{# nonGNU #}` to c2hs file. There are 4 chs files:
```
./src/Foreign/R/Error.chs
./src/Foreign/R/Embedded.chs
./src/Foreign/R/Parse.chs
./src/Foreign/R.chs
```
Can you try those fixes for you?  Sorry for that troubles, but unfortunately I don't have a system where this problem is reproducible, to test fix. And travis-ci with macos that we are using is completely fine.

**Comment by almays** [comment]
My difference from travis is `gcc` version:

    Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
    Apple LLVM version 8.0.0 (clang-800.0.42.1)
    Target: x86_64-apple-darwin16.1.0
    Thread model: posix
    InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

And in travis there is:

    Apple LLVM version 7.3.0 (clang-703.0.31)
    Target: x86_64-apple-darwin15.6.0

**Comment by almays** [comment]
I tried `--docker`. Well, there it builds. But there when I run `ipython notebook` it says, that browser not found.

I surrender. I have in my native `jupyter` both `IHaskell` and `R` separately. So it's enough for my experiments for now.

**Comment by mboes** [comment]
@almays does adding `{# nonGNU #}` to the top of all the .chs files as @qnikst suggests above work for you? If so we can push such a change to master to have this fixed for others too.

**Comment by almays** [comment]
No. Adding `{# nonGNU #}` doesn't help. I also tried to make a `ln -s` to brew installed `gcc` in PATH earlier, than apple one. No effect.

As you see earlier in the `stack build` output, if I run it consequently, then I get two times errors about "symbol does not fit here". Now in both errors like that I also get new things:

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/Dev/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Configuring inline-r-0.9.0.0...
    Preprocessing library inline-r-0.9.0.0...
    In file included from .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/Foreign/R.chs.h:12:
    In file included from /usr/local/Cellar/r/3.3.2/R.framework/Versions/3.3/Resources/include/R.h:45:
    In file included from /usr/include/stdlib.h:63:
    In file included from /usr/include/_types.h:27:
    In file included from /usr/include/sys/_types.h:32:
    /usr/include/sys/cdefs.h:81:2: warning: "Unsupported compiler detected" [-W#warnings]
    #warning "Unsupported compiler detected"
     ^
    1 warning generated.
    c2hs: C header contains errors:

    /usr/include/stdlib.h:139: (column 38) [ERROR]  >>> Syntax error !
      The symbol `__compar' does not fit here.


    Dahan-Macbook:HaskellR dahan$ stack --no-terminal -j2 build --pedantic
    No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
    Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
    This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595

    inline-r-0.9.0.0: build

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/Dev/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Preprocessing library inline-r-0.9.0.0...
    In file included from .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/Foreign/R/Embedded.chs.h:3:
    In file included from cbits/missing_r.h:6:
    In file included from /Users/dahan/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/include/HsFFI.h:30:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.0.0/include/stdint.h:63:
    In file included from /usr/include/stdint.h:52:
    In file included from /usr/include/sys/_types.h:32:
    /usr/include/sys/cdefs.h:81:2: warning: "Unsupported compiler detected" [-W#warnings]
    #warning "Unsupported compiler detected"
     ^
    1 warning generated.
    c2hs: C header contains errors:

    /usr/include/stdio.h:133: (column 19) [ERROR]  >>> Syntax error !
      The symbol `_close' does not fit here.


    Dahan-Macbook:HaskellR dahan$ stack --no-terminal -j2 build --pedantic
    No packages found in snapshot which provide a "hsc2hs" executable, which is a build-tool dependency of "inline-r"
    Missing build-tools may be caused by dependencies of the build-tool being overridden by extra-deps.
    This should be fixed soon - see this issue https://github.com/commercialhaskell/stack/issues/595

    inline-r-0.9.0.0: build

    --  While building package inline-r-0.9.0.0 using:
      /Users/dahan/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build lib:inline-r --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/dahan/Dev/HaskellR/.stack-work/logs/inline-r-0.9.0.0.log

    Preprocessing library inline-r-0.9.0.0...
    c2hs: Foreign/R.chi not found in:
    .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build
    .

`ihaskell` gets built by `stack --no-terminal -j2 build --only-snapshot --prefetch --test --bench`. And it can install it's kernel to Jupyter. But it is a plain normal Haskell kernel. All `[r|]` syntax is not supported.

**Comment by biglambda** [comment]
I get the same exact error attempting to build [](https://github.com/IFCA/opencl )after I upgraded to OSX Sierra. c2hs is stopped by this line:
```
c2hs: C header contains errors:

    /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/stdlib.h:139: (column 38) [ERROR]  >>> Syntax error !
      The symbol `__compar' does not fit here.
```

No special stack options in my case.

**Comment by deech** [comment]
Hi,
I was able to work around these issues in my C2HS project: https://github.com/haskell/c2hs/issues/176#issuecomment-279163958.

**Comment by qnikst** [comment]
This should be fixed with 0.9.1 release. 
Feel free to reopen if problem still appear after that.

**Comment by Fizzixnerd** [comment]
I am having this problem too, though it's on Linux not macOS.

**Comment by qnikst** [comment]
@Fizzixnerd do you use latest released inline-r? (0.9.1)
